File: Freelance/WhatsSUP/WhatsSUP1/WhatsSUP1/SavedPathLocations.m
//
//  SavedPathLocations.m
//  WhatsSUP1
//
//  Created by Tim Teece on 03/07/2014.
//  Copyright (c) 2014 Tim Teece. All rights reserved.
//

#import "SavedPathLocations.h"
#import "SavedSUPPath.h"


@implementation SavedPathLocations

@dynamic altitude;
@dynamic degrees;
@dynamic direction;
@dynamic hAcc;
@dynamic lat;
@dynamic lon;
@dynamic speed;
@dynamic stepNo;
@dynamic timestamp;
@dynamic vAcc;
@dynamic inPath;
@dynamic totDistance;
@dynamic cloudKitId;

-(CKToCDMap*) saveToCloudKit:(CKDatabase *)db addToPath:(CKRecordID*)pthid{
    
    if ([self.cloudKitId length] ==0){
        CKRecord * pathRec= [[CKRecord alloc ] initWithRecordType:@"SavedSUPPathLocation"];
        
        [pathRec setObject:self.altitude forKey:@"altitude"];
        [pathRec setObject:self.degrees forKey:@"degrees"];
        [pathRec setObject:self.direction forKey:@"direction"];
        [pathRec setObject:self.hAcc forKey:@"hAcc"];
        
        CLLocation *end = [[CLLocation alloc] initWithLatitude:self.lat.doubleValue longitude:self.lon.doubleValue];
        [pathRec setObject:end forKey:@"loc"];
        
        [pathRec setObject:self.speed forKey:@"speed"];
        [pathRec setObject:self.stepNo forKey:@"stepNo"];
        [pathRec setObject:self.timestamp forKey:@"timestamp"];
        [pathRec setObject:self.vAcc forKey:@"vAcc"];
        [pathRec setObject:self.totDistance forKey:@"totDistance"];
        //CKReference * pathReference=[[CKReference alloc] initWithRecord:pth action:CKReferenceActionNone];
        
        CKReference * pathReference=[[CKReference alloc] initWithRecordID:pthid action:CKReferenceActionNone];
        pathRec[@"PathRef"]=pathReference;
        
        self.cloudKitId=pathRec.recordID.description;

        /*
        [db saveRecord:pathRec completionHandler:^(CKRecord *savedPath,NSError* error){
            if(error!=nil){
                NSLog(@"Something went wrong");
                NSLog(@"%@",error.description);
            }else{
                // save the path values
                NSLog(@"%@",savedPath.recordID.description);
                // link this to the path
                
                self.cloudKitId=savedPath.recordID.description;
                
                @try{
                    NSError* error1=nil;
                    [[self managedObjectContext] save:&error1];
                }@catch(NSException *e){
                    NSLog(@"Error saving to cloud %@", [e description]);
                }
                
            }
        }];*/
        CKToCDMap * ret= [[CKToCDMap alloc] init];
        [ret setCdRef:self];
        [ret setCkRef:pathRec];
        return ret;
    }else{
        return nil;
    }
   
}



@end

File: Freelance/WhatsSUP/WhatsSUP1/WhatsSUP1/SavedPathLocations.swift
//
//  SavedPathLocations.swift
//  WhatsSUP1
//
//  Created by Tim Teece on 01/07/2014.
//  Copyright (c) 2014 Tim Teece. All rights reserved.
//

import Foundation
import UIKit
import CoreData
import CoreLocation

enum SavedPathLocationsPropertyList {
    case altitude,degrees,direction,hAcc,lat,lon,speed,stepNo,timestamp,vAcc,inPath
    func description() -> String {
        switch self {
        case .altitude:
            return "altitude"
        case .degrees:
            return "degrees"
        case .direction:
            return "direction"
        case .hAcc:
            return "hAcc"
        case .lat:
            return "lat"
        case .lon:
            return "lon"
        case .speed:
            return "speed"
        case .stepNo:
            return "stepNo"
        case .timestamp:
            return "timestamp"
        case .vAcc:
            return "vAcc"
        case .inPath:
            return "inPath"
        }
    }
}

@objc(SavedPathLocations)
class SavedPathLocations: NSManagedObject {
    
    @NSManaged var altitude: Double
    @NSManaged var degrees: Double
    @NSManaged var direction: Double
    @NSManaged var hAcc: Double
    @NSManaged var lat: Double
    @NSManaged var lon: Double
    @NSManaged var speed: Double
    @NSManaged var stepNo: Int64
    @NSManaged var timestamp: NSDate
    @NSManaged var vAcc: Double
    @NSManaged var inPath: SavedSUPPath?
    
    
    //
    //// CREATE CLASS OBJECT
    //
    
    
    class func createMyObject (location:CLLocation, count: Int64, context: NSManagedObjectContext) -> SavedPathLocations? {
        //   let propertyType = propertyName.description()
        let entityName = "SavedPathLocations"
        let entityDescription = NSEntityDescription.entityForName(entityName, inManagedObjectContext: context)
        var myObject : SavedPathLocations = SavedPathLocations(entity: entityDescription, insertIntoManagedObjectContext: context)
        println(location.altitude)
        
        myObject.stepNo=count
        myObject.direction=location.course
        myObject.hAcc=location.horizontalAccuracy
        myObject.vAcc=location.verticalAccuracy
        if (myObject.vAcc > 0 && myObject.vAcc < 20){
            myObject.altitude=location.altitude
        }
        if (myObject.hAcc > 0 && myObject.hAcc < 20){
            myObject.lat=location.coordinate.longitude
            myObject.lon=location.coordinate.longitude
        }
        myObject.speed=location.speed
        myObject.timestamp=location.timestamp
        var error: NSError? = nil
        context.save(&error)
        println("ID of SavedPathLocations : \(myObject.objectID)")
        return myObject
    }
    
    
/*
    
    class func createMyObject (location:CLLocation, context: NSManagedObjectContext) -> SavedPathLocations? {
     //   let propertyType = propertyName.description()
        
        let entityName = "SavedPathLocations"
        let request : NSFetchRequest = NSFetchRequest(entityName: entityName)
        request.returnsObjectsAsFaults = false
      //  request.predicate = NSPredicate(format: "\(propertyType) = %@", value)
        var error: NSError? = nil
        var matches: NSArray = context.executeFetchRequest(request, error: &error)
        
        if (matches.count > 1) {
            // handle error
            return matches[0] as? SavedPathLocations
        } else if matches.count ==  0 {
            let entityDescription = NSEntityDescription.entityForName(entityName, inManagedObjectContext: context)
            var myObject : SavedPathLocations = SavedPathLocations(entity: entityDescription, insertIntoManagedObjectContext: context)
            myObject.name = value
            return myObject
        }
        else {
            println(matches[0])
            return matches[0] as? SavedPathLocations
        }
 
    }*/
}
